### DISCUSSION TITLE
유니티 프로젝트에서는 레이캐스트가 작동하는데, 모바일에서 작동하지 않습니다.

### DISCUSSION THREAD

**Discussion by ryusiin** [naverz/zepeto-studio-kor#527] [5. Q&A]
원인을 아시는 분 있을까요?

유니티 에디터
![image](https://user-images.githubusercontent.com/5895396/159844470-470f6518-4c1b-44d8-955c-b2493a589833.png)

모바일
![image](https://user-images.githubusercontent.com/5895396/159844479-26251ac6-2710-48c6-96f5-6c84daa989b2.png)

// :: 코드 Code
`private * ShootRay() {
        while (true) {
            // :: 스타팅 포지션
            const startingPosition = new UnityEngine.Vector3(
                this.gameObject.transform.position.x,
                this.gameObject.transform.position.y + 0.5,
                this.gameObject.transform.position.z
            );

            // :: 레이 제작
            let ray: UnityEngine.Ray = new UnityEngine.Ray(
                startingPosition, this.gameObject.transform.forward);

            // RaycatHit 참조변수를 파라메터로 넘기기 위해, $ref 연산자로 wrapping 하는 부분에 유의해주세요.
            // (매개변수 한정자 out이 필요한 상황이기때문에) 객체 생성 없이 참조만 선언해줍니다.            
            let ref = $ref<UnityEngine.RaycastHit>();

            // :: 레이어 마스크
            const layerMask = 7 << UnityEngine.LayerMask.NameToLayer("NPC");

            if (UnityEngine.Physics.Raycast(ray, ref, 1000, layerMask)) {
                // RaycatHit 데이터를 확인하기 위해, $unref 연산자로 다시 unwrapping 하는 부분에 유의해주세요.
                let hitInfo = $unref(ref);

                // :: NPC Data 획득
                const npcData = hitInfo.collider.gameObject.GetComponent<NPCData>();
                if (npcData != null) {
                    if (this.iCurTarget_NPC != null) {
                        this.iCurTarget_NPC.HideTargetCube();
                    }

                    this.iCurTarget_NPC = npcData;
                    this.iCurTarget_NPC.ShowTargetCube();
                }
            } else {
                if (this.iCurTarget_NPC != null) {
                    this.iCurTarget_NPC.HideTargetCube();
                    this.iCurTarget_NPC = null;
                }

                UnityEngine.Debug.DrawRay(
                    startingPosition, this.gameObject.transform.forward,
                    UnityEngine.Color.red, 0.1);
            }

            yield new UnityEngine.WaitForSeconds(0.1);
        }
    }`

**Comment by x4cf** [comment]
레이어 번호를 확인해보는게 좋을것같아요. 제페토에서는 사용 가능한 레이어 번호가 정해져있습니다 :D 

https://studio.zepeto.me/guides/tag-and-layer
<img width="678" alt="레이어" src="https://user-images.githubusercontent.com/90679926/159847385-821430c7-7583-4008-8917-fcfcdb44acdc.png">

**Reply by ryusiin** [reply]
레이어 번호를 바꿔 봤는데도, 여전히 안되네요...

![image](https://user-images.githubusercontent.com/5895396/159849715-012e65c6-68dd-4e3e-9762-dff32f6b0fe4.png)

**Reply by x4cf** [reply]
코드중 레이어 변수 설정에 문제를 발견하였습니다 
레이어 충돌 체크를 이렇게 해보는걸 추천합니다.

let ray = Camera.main.ScreenPointToRay(Input.mousePosition);
let ref = $ref<RaycastHit>();    
              
let layerMask = (1 << LayerMask.NameToLayer("TestLayerA")) + (1 << LayerMask.NameToLayer("TestLayerB"));
 
if(Physics.Raycast(ray, ref, 1000, layerMask)){
let hitInfo = $unref(ref);
 
if((hitInfo.transform.gameObject.layer == LayerMask.NameToLayer("TestLayerA")) ||
  (hitInfo.transform.gameObject.layer == LayerMask.NameToLayer("TestLayerB"))){
// 충돌됨
}
}

**Reply by x4cf** [reply]
이 경우 정상적인 레이어 마스크 넘버를 반환하지 않습니다
const layerMask = 7 << UnityEngine.LayerMask.NameToLayer("NPC");

**Reply by ryusiin** [reply]
감사합니다:>!

레이캐스트 관련 문제는 해결했습니다.

뭔가 이상해서 체커를 달아봤는데,
레이캐스트가 아니라 모바일에서는 큐브가 보여지지 않는 게 문제인 모양입니다...

![image](https://user-images.githubusercontent.com/5895396/159855423-36a5c644-c6eb-4452-b805-be867b494986.png)

확인 코드입니다

private * ShootRay() {
        while (true) {
            // :: 스타팅 포지션
            const startingPosition = new UnityEngine.Vector3(
                this.gameObject.transform.position.x,
                this.gameObject.transform.position.y + 0.5,
                this.gameObject.transform.position.z
            );

            // :: 레이 제작
            let ray: UnityEngine.Ray = new UnityEngine.Ray(
                startingPosition, this.gameObject.transform.forward);

            // RaycatHit 참조변수를 파라메터로 넘기기 위해, $ref 연산자로 wrapping 하는 부분에 유의해주세요.
            // (매개변수 한정자 out이 필요한 상황이기때문에) 객체 생성 없이 참조만 선언해줍니다.            
            let ref = $ref<UnityEngine.RaycastHit>();

            // :: 레이어 마스크
            const layerMask = (1 << UnityEngine.LayerMask.NameToLayer("NPC"))
                + (1 << UnityEngine.LayerMask.NameToLayer("Player"));

            if (UnityEngine.Physics.Raycast(ray, ref, 1000, layerMask)) {
                // RaycatHit 데이터를 확인하기 위해, $unref 연산자로 다시 unwrapping 하는 부분에 유의해주세요.
                let hitInfo = $unref(ref);

                // :: 충돌 확인
                PastaRuler.oInstance.UIManager.CheckCollider(true);

                // :: NPC Data 획득
                const npcData = hitInfo.collider.gameObject.GetComponent<NPCData>();
                if (npcData != null) {
                    if (this.iCurTarget_NPC != null) {
                        this.iCurTarget_NPC.HideTargetCube();
                    }

                    this.iCurTarget_NPC = npcData;
                    this.iCurTarget_NPC.ShowTargetCube();
                }
            } else {
                // :: 충돌 확인
                PastaRuler.oInstance.UIManager.CheckCollider(false);

                if (this.iCurTarget_NPC != null) {
                    this.iCurTarget_NPC.HideTargetCube();
                    this.iCurTarget_NPC = null;
                }

                UnityEngine.Debug.DrawRay(
                    startingPosition, this.gameObject.transform.forward,
                    UnityEngine.Color.red, 0.1);
            }

            yield new UnityEngine.WaitForSeconds(0.1);
        }
    }
